src/Rules.py

```diff
@@ -78,27 +78,39 @@ def checkRequireStringForArea(state: CollectionState, area: dict):
-            func_args = item[1].split(",")
-            if func_args == ['']:
-                func_args.pop()
+        def findAndRecursivelyExecuteFunctions(requires_list: str, maxRecursion: int, recursionDepth: int = 0) -> str:
+            found_functions = re.findall(r'\{(\w+)\((.*?)\)\}', requires_list)
+            if found_functions:
+                if recursionDepth >= maxRecursion:
+                    raise RecursionError(f'One or more functions in "{area.get("name", f"An area with these parameters: {area}")}"\'s requires looped too many time (maximum recursion is {maxRecursion}) \
+                                         \n    As of this Exception the following function(s) are waiting to run: {[f[0] for f in found_functions]} \
+                                         \n    And the currently processed requires look like this: "{requires_list}"')
+                else:
+                    for item in found_functions:
+                        func_name = item[0]
+                        func_args = item[1].split(",")
+                        if func_args == ['']:
+                            func_args.pop()
 
-            func = globals().get(func_name)
+                        func = globals().get(func_name)
 
-            if func is None:
-                func = getattr(Rules, func_name, None)
+                        if func is None:
+                            func = getattr(Rules, func_name, None)
 
-            if not callable(func):
-                raise ValueError(f"Invalid function `{func_name}` in {area}.")
+                        if not callable(func):
+                            raise ValueError(f"Invalid function `{func_name}` in {area}.")
 
-            convert_req_function_args(func, func_args, area.get("name", f"An area with these parameters: {area}"))
-            result = func(world, multiworld, state, player, *func_args)
-            if isinstance(result, bool):
-                requires_list = requires_list.replace("{" + func_name + "(" + item[1] + ")}", "1" if result else "0")
-            else:
-                requires_list = requires_list.replace("{" + func_name + "(" + item[1] + ")}", str(result))
+                        convert_req_function_args(func, func_args, area.get("name", f"An area with these parameters: {area}"))
+                        result = func(world, multiworld, state, player, *func_args)
+                        if isinstance(result, bool):
+                            requires_list = requires_list.replace("{" + func_name + "(" + item[1] + ")}", "1" if result else "0")
+                        else:
+                            requires_list = requires_list.replace("{" + func_name + "(" + item[1] + ")}", str(result))
+
+                requires_list = findAndRecursivelyExecuteFunctions(requires_list, maxRecursion, recursionDepth + 1)
+            return requires_list
 
+        requires_list = findAndRecursivelyExecuteFunctions(requires_list, maxRecursion=5)
```

Comment:
Probably better to hardcode max recursion instead of passing it in, either in the function or on the world. That way, anyone reusing this later can't accidentally screw up their max recursion and introduce an infinite loop.

I doubt we want a customizable max per run, anyways.

Reply:
ill put the variable on the world so dev can change it via hooks if they really wants even if more recursion = more generation time